packages/core/src/electron-browser/menu/electron-main-menu-factory.ts

```diff
@@ -71,33 +71,49 @@ export class ElectronMainMenuFactory {
         @inject(MenuModelRegistry) protected readonly menuProvider: MenuModelRegistry,
         @inject(KeybindingRegistry) protected readonly keybindingRegistry: KeybindingRegistry
     ) {
-        preferencesService.onPreferenceChanged(debounce(() => {
-            if (this._menu) {
-                for (const item of this._toggledCommands) {
-                    this._menu.getMenuItemById(item).checked = this.commandRegistry.isToggled(item);
+        preferencesService.onPreferenceChanged(
+            debounce(e => {
+                if (e.preferenceName === 'window.menuBarVisibility') {
+                    this.setMenuBar();
                 }
-                electron.remote.getCurrentWindow().setMenu(this._menu);
-            }
-        }, 10));
+                if (this._menu) {
+                    for (const item of this._toggledCommands) {
+                        this._menu.getMenuItemById(item).checked = this.commandRegistry.isToggled(item);
+                    }
+                    electron.remote.getCurrentWindow().setMenu(this._menu);
+                }
+            }, 10)
+        );
         keybindingRegistry.onKeybindingsChanged(() => {
-            const createdMenuBar = this.createMenuBar();
-            if (isOSX) {
-                electron.remote.Menu.setApplicationMenu(createdMenuBar);
-            } else {
-                electron.remote.getCurrentWindow().setMenu(createdMenuBar);
-            }
+            this.setMenuBar();
         });
     }
 
-    createMenuBar(): Electron.Menu {
-        const menuModel = this.menuProvider.getMenu(MAIN_MENU_BAR);
-        const template = this.fillMenuTemplate([], menuModel);
+    setMenuBar(): void {
+        const createdMenuBar = this.createMenuBar();
         if (isOSX) {
-            template.unshift(this.createOSXMenu());
+            electron.remote.Menu.setApplicationMenu(createdMenuBar);
+        } else {
+            electron.remote.getCurrentWindow().setMenu(createdMenuBar);
+        }
+    }
+
+    createMenuBar(): Electron.Menu | null {
+        let preference = this.preferencesService.get<string>('window.menuBarVisibility');
+        preference = preference ? preference : 'classic';
+        if (preference && ['classic', 'visible'].includes(preference)) {
+            const menuModel = this.menuProvider.getMenu(MAIN_MENU_BAR);
+            const template = this.fillMenuTemplate([], menuModel);
+            if (isOSX) {
+                template.unshift(this.createOSXMenu());
+            }
+            const menu = electron.remote.Menu.buildFromTemplate(template);
+            this._menu = menu;
+            return this._menu;
         }
-        const menu = electron.remote.Menu.buildFromTemplate(template);
-        this._menu = menu;
-        return menu;
+        this._menu = undefined;
+        // eslint-disable-next-line no-null/no-null
+        return null;
```

Comment:
We should avoid using `null` and use `undefined` instead. Same in other areas.

Reply:
I looked and saw the menu in electron accepts types : electron.Menu | null (and not undefined)
`    private setMenu(menu: electron.Menu | null = this.factory.createMenuBar(), electronWindow: electron.BrowserWindow = electron.remote.getCurrentWindow()): void {
        if (isOSX) {
            electron.remote.Menu.setApplicationMenu(menu);
        } else {
            // Unix/Windows: Set the per-window menus
            electronWindow.setMenu(menu);
        }
    }`
we can change the return value of the function, but then we will have to handle the undefined value and replace with null when using the menu functions. (there are few more places calling the functions after create)

should we do that?

Reply:
I see, we can accept `null` then:

- https://github.com/electron/electron/blob/main/docs/api/browser-window.md#winsetmenumenu-linux-windows
- https://github.com/electron/electron/blob/main/docs/api/browser-window.md#winremovemenu-linux-windows
 
We'll need to confirm this actually works on `macOS`.